src/native/eventpipe/ep.c

```diff
@@ -1313,15 +1317,25 @@ ep_delete_provider (EventPipeProvider *provider)
 	// Take the lock to make sure that we don't have a race
 	// between disabling tracing and deleting a provider
 	// where we hold a provider after tracing has been disabled.
+	bool wait_for_provider_callbacks_completion = false;
 	EP_LOCK_ENTER (section1)
 		if (enabled ()) {
 			// Save the provider until the end of the tracing session.
 			ep_provider_set_delete_deferred (provider, true);
+
+			// the callback func must be previously set to null,
+			// otherwise callbacks might never stop coming
+			EP_ASSERT (provider->callback_func == NULL);
+			if (provider->uninvoked_prepared_callbacks_counter > 0)
+				wait_for_provider_callbacks_completion = true;
```

Comment:
I'd add a comment on EventPipeEventProvider.Unregister() and here in ep_delete_provider() that if these functions are called within a callback the current behavior is a deadlock. That deadlock condition already exists for ETW and we have [an assert](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventProvider.cs,165) that would fire in debug builds, but nothing actually prevents users from writing their code to hit it. Its possible we'll see someone who has written an EventSource which does this and they never encountered the deadlock before because they always used EventPipe instead of ETW to do their tracing.

I'd recommend we open a bug describing that issue but I'd be fine if it gets fixed separately as the ETW portion is a pre-existing problem. A repro is roughly:

```C#
MyEventSource eventSource = new MyEventSource();
Console.WriteLine("MyEventSource is read to be enabled");
Console.ReadLine();

[EventSource(Name ="MyEventSource")]
class MyEventSource : EventSource
{
    override protected void OnEventCommand(EventCommandEventArgs command)
    {
        Console.WriteLine($"Command: {command.Command}");
        this.Dispose();
        Console.WriteLine($"Dispose() complete");
    }
}
```

If you use ETW to enable the provider you will see 'Command' get printed, but not 'Dispose()' because the thread hangs. After this PR the same behavior will also be true for EventPipe.

Reply:
Added the comment and [created an issue for the deadlock](https://github.com/dotnet/runtime/issues/106087)